module/userExtension/helper.js

```diff
@@ -211,304 +212,373 @@ module.exports = class UserExtensionHelper {
+                          });
+                        }
+                      }
+                    } else {
+                      // Create new program entry
+                      existingUserPlatformRoles.push({
+                        programId: programObjectId,
+                        roles: [...newRoles],
+                      });
+
+                      // All roles are new, emit create events
+                      for (const role of newRoles) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programObjectId,
+                          role,
+                          eventType: 'create',
+                        });
+                      }
+                    }
+                  } else if (userRole.programOperation === 'REMOVE') {
+                    if (currentRoleInfoIndex !== -1) {
+                      const currentRoles = existingUserPlatformRoles[currentRoleInfoIndex].roles;
+
+                      // Remove specified roles
+                      const rolesToKeep = currentRoles.filter((role) => !newRoles.includes(role));
+
+                      // Emit delete events for removed roles
+                      const rolesToRemove = currentRoles.filter((role) => newRoles.includes(role));
+                      for (const role of rolesToRemove) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programObjectId,
+                          role,
+                          eventType: 'delete',
+                        });
+                      }
+
+                      if (rolesToKeep.length === 0) {
+                        // Remove entire program entry if no roles left
+                        existingUserPlatformRoles.splice(currentRoleInfoIndex, 1);
+                      } else {
+                        // Update with remaining roles
+                        existingUserPlatformRoles[currentRoleInfoIndex].roles = rolesToKeep;
+                      }
+                    }
+                    // If program doesn't exist, nothing to remove - no error needed
                   }
                 }
+              }
 
-                existingUser.platformRoles[userPlatformRoleToUpdate].programs = existingUser.platformRoles[
-                  userPlatformRoleToUpdate
-                ].programs.map((eachProgram) => ObjectId(eachProgram));
+              // Update user extension document
+              const updateQuery = { platformRoles: existingUserPlatformRoles };
 
-                updateQuery['platformRoles'] = existingUser.platformRoles;
-              }
+              user = await database.models.userExtension.findOneAndUpdate({ _id: existingUser._id }, updateQuery, {
```

Comment:
@borkarsaish65 Why are we directly calling db creation/delete/update from helper? Don't we have a directory called databaseQueries and from there we channel all db interactions right? Please check this once.